storage-proofs/porep/src/stacked/vanilla/memory_handling.rs

```diff
@@ -0,0 +1,255 @@
+use std::cell::UnsafeCell;
+use std::fs::File;
+use std::marker::{PhantomData, Sync};
+use std::mem::size_of;
+use std::path::PathBuf;
+
+use anyhow::Result;
+use byte_slice_cast::*;
+use log::*;
+use mapr::{Mmap, MmapMut, MmapOptions};
+use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering::SeqCst};
+
+pub struct CacheReader<T> {
+    file: File,
+    bufs: UnsafeCell<[Mmap; 2]>,
+    size: usize,
+    degree: usize,
+    window_size: usize,
+    cur_window: AtomicUsize,
+    cur_window_safe: AtomicUsize,
+    _t: PhantomData<T>,
+}
+
+unsafe impl<T> Sync for CacheReader<T> {}
+
+impl<T: FromByteSlice> CacheReader<T> {
+    pub fn new(filename: &PathBuf, window_size: Option<usize>, degree: usize) -> Result<Self> {
+        info!("initializing cache");
+        let file = File::open(filename)?;
+        let size = File::metadata(&file)?.len() as usize;
+        let window_size = match window_size {
+            Some(s) => {
+                if s < size {
+                    assert_eq!(
+                        0,
+                        size % degree * size_of::<T>(),
+                        "window size is not multiple of element size"
+                    );
+                };
+                s
+            }
+            None => {
+                let num_windows = 8;
+                assert_eq!(0, size % num_windows);
+                size / num_windows
+            }
+        };
+
+        let buf0 = Self::map_buf(0, window_size, &file)?;
+        let buf1 = Self::map_buf(window_size as u64, window_size, &file)?;
+        Ok(Self {
+            file,
+            bufs: UnsafeCell::new([buf0, buf1]),
+            size,
+            degree,
+            window_size,
+            // The furthest window from which the cache has yet been read.
+            cur_window: AtomicUsize::new(0),
+            cur_window_safe: AtomicUsize::new(0),
+            _t: PhantomData::<T>,
+        })
+    }
+
+    #[inline]
+    fn get_bufs(&self) -> &[Mmap] {
+        unsafe { &std::slice::from_raw_parts((*self.bufs.get()).as_ptr(), 2) }
+    }
+
+    #[inline]
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn get_mut_bufs(&self) -> &mut [Mmap] {
+        std::slice::from_raw_parts_mut((*self.bufs.get()).as_mut_ptr(), 2)
+    }
+
+    // TODO: is this actually needed?
```

Comment:
It's not currently used. It was split into the start and finish portions and split across layer-transition boundaries. I think it remains useful documentation of intent. It could probably be refactored into a call to start then finish, which would serve that purpose even better.